src/coins.cpp

```diff
@@ -1203,14 +1217,20 @@ bool CCoinsViewCache::IsTxCswApplicableToState(const CTransaction& tx, CValidati
             }
         }
 
-        const libzendoomc::ScFieldElement& prevCumulativeCertDataHash = libzendoomc::ScFieldElement(); // GetCumulativeCertDataHash(csw.scId, csw.nEpoch - 1);
-        const libzendoomc::ScFieldElement& currentCertDataHash = libzendoomc::ScFieldElement(); // GetCertDataHash(csw.scId, csw.nEpoch);
-        const libzendoomc::ScFieldElement& lastCumulativeCertDataHash = libzendoomc::ScFieldElement(); // GetLastCumulativeCertDataHash(csw.scId);
+        std::pair<libzendoomc::ScFieldElement, libzendoomc::ScFieldElement> certDataHashes;
+        if (!this->GetCertDataHashes(csw.scId, csw.nEpoch, certDataHashes))
+            return state.Invalid(error("missing cert data hash for required scId/epoch"),
+                 REJECT_INVALID, "tx-csw-missing-cert-data-hash");
+
+        const libzendoomc::ScFieldElement& prevCumulativeCertDataHash = certDataHashes.second;
+        const libzendoomc::ScFieldElement& currentCertDataHash = certDataHashes.first;
+        libzendoomc::ScFieldElement lastCumulativeCertDataHash = libzendoomc::ScFieldElement();
```

Comment:
We need to get hashes for the most recent epoch valid Certificate -> then calculate the cumulative one.
For example, SC ceased at epoch 100, so the last Certificate was submitted in epoch 99, but will be invalidated.
So, the most recent one was submitted in epoch 98 for epoch 97.
We need to call `GetCertDataHashes(csw.scId, 97, recentDataHashes)`
I guess that `97` is taken from CSidechain entry for given `scId`.

Reply:
Done